Add with clause (#9)
aleDsz/elixir4noobs: contents/3 - Linguagem/7-Estruturas de Decisao.md

...>   true ->
...>     "acho que não teve jeito mesmo"
...> end
```

### With

> O with é como um "execute se receber tal", dado uma função, se o seu retorno for o esperado, execute o código. Ficara mais facil entender com o exemplo a seguir

Reviewer: Acho que aqui seria bom referenciar que é possível realizar o mesmo conceito de `pattern matching`.

> O with é como um "execute se receber tal", dado uma função, se o seu retorno for o esperado, execute o código. Ficara mais facil entender com o exemplo a seguir

Reviewer: ```suggestion
O `with` é como um "execute se receber tal", dado uma função, se o seu retorno for o esperado, execute o código. Ficará mais fácil entender com o exemplo a seguir:
```

```elixir
iex> with {:ok, _banana} <- MyModule.my_fun(my_param) do
...>  "Deu boa patrão"
...> end
```

> Basicamente se a função retornar uma tupla de `{:ok, qualquer_valor}` o treixo de código no bloco sera executado, caso contrario um erro é retornado pra quem chamou a função.

Reviewer: ```suggestion
Basicamente se a função retornar uma tupla de `{:ok, qualquer_valor}` o trecho de código no bloco será executado. Caso contrario, um erro é retornado pra quem chamou a função.
```

Reviewer: Aqui também, acho que da pra elaborar um pouco mais, utilizando uma escrita mais simples para que não haja dúvidas sobre o seu uso.

> O with pode receber varias condições para avaliar, e também pode conter um `else` para tratar os erros.

Reviewer: ```suggestion
O `with` pode receber varias condições para avaliar, e também pode conter um bloco de `else` para tratar qualquer clausula que não seja esperada no bloco `with ... do`.
```

...>   true ->
...>     "acho que não teve jeito mesmo"
...> end
```

### With

> Podia ser um `with` ela (ou ele, ou elu), mas é só uma função do `Kernel` para avaliar uma ou mais condições, é recomendado utilizar em momentos que seu `case` começar a ficar aninhado.

Reviewer: ```suggestion
É uma função do `Kernel` para avaliar uma ou mais condições, é recomendado utilizar em momentos que seu fluxo de execução requerir validação de [pattern matching](../4%20-%20Recursos%20extras/3-Pattern%20Matching.md) de inúmeras condições onde o `case` não seria o suficiente, principalmente quando você não precisa lidar com os resultados não esperados.
```

> Dado uma função, se o seu retorno for o esperado (para isso sera utilizado o [pattern matching](../4%20-%20Recursos%20extras/3-Pattern%20Matching.md)), execute o código. Ficará mais fácil entender com o exemplo a seguir:

Reviewer: ```suggestion
Dado uma função, se o seu retorno for o esperado (para isso será utilizado o [pattern matching](../4%20-%20Recursos%20extras/3-Pattern%20Matching.md)), execute o código. Ficará mais fácil entender com o exemplo a seguir:
```

```elixir
iex> with {:ok, _banana} <- MyModule.my_fun(my_param) do
...>  "Deu boa patrão"
...> end
```

Basicamente, se a função retornar uma tupla de `{:ok, qualquer_valor}` (valores que você pode definir como quiser), o trecho de código no bloco `do` será executado. Caso contrario, um erro é retornado para quem chamou a função.

O `with` pode receber várias condições para avaliar, e também pode conter um bloco de `else` para tratar qualquer cláusula que não seja esperada no bloco `with ... do`. Para tratar tais cláusulas você pode usar uma espécie de `case`.

```elixir
iex> with {:ok, _banana} <- MyModule.my_fun(my_param),
...>      _another_banana <- MyModule.my_other_fun(my_param) do
...>  "Deu boa patrão"
...> else
...>  :error -> 
...>    "Deu merda patrão"
...>
...>  _other_error ->
...>    "Deu outra merda patrão"
...> end
```

> 

Reviewer: ```suggestion

Um outro bom exemplo de uso do `with ... do` é quando você não precisa tratar os valores não esperados que você teria de tratar com o `case .. do`:

[code]
# Case
case Modulo.funcao() do
  {:ok, _banana} -> "banana"
  {:error, _motivo} -> "maçã"
  _ -> "limão"
end

# With
with {:ok, _banana} <- Modulo.funcao() do
  "banana"
end
[/code]

Quando retornar `{:error, _}` ou qualquer outro resultado, ele pode lidar isso em outro ponto do fluxo da execução, não necessitando de lidar com todos os retornos em todas as funções, mantendo tudo isso direto em apenas um único local.

```

Reviewer: Deixei o `[code]` pois não da pra colocar o bloco de código sem quebrar o bloco de `suggestion` do Github
